server/src/test/java/org/opensearch/action/admin/cluster/node/tasks/ResourceAwareTasksTests.java

```diff
@@ -0,0 +1,633 @@
+        setup(true, false);
+
+        final AtomicReference<Throwable> throwableReference = new AtomicReference<>();
+        final AtomicReference<NodesResponse> responseReference = new AtomicReference<>();
+        TaskTestContext taskTestContext = new TaskTestContext();
+
+        Map<Long, Task> resourceTasks = testNodes[0].taskResourceTrackingService.getResourceAwareTasks();
+
+        taskTestContext.operationStartValidator = threadId -> {
+            Task task = resourceTasks.values().stream().findAny().get();
+            // One thread is currently working on task but not finished
+            assertEquals(1, resourceTasks.size());
+            assertEquals(1, task.getResourceStats().size());
+            assertEquals(1, task.getResourceStats().get(threadId).size());
+            assertTrue(task.getResourceStats().get(threadId).get(0).isActive());
+            assertEquals(0, task.getTotalResourceStats().getCpuTimeInNanos());
+            assertEquals(0, task.getTotalResourceStats().getMemoryInBytes());
+
+            testNodes[0].taskResourceTrackingService.setTaskResourceTrackingEnabled(false);
+        };
+
+        taskTestContext.operationFinishedValidator = threadId -> {
+            Task task = resourceTasks.values().stream().findAny().get();
+            // Thread has finished working on the task's runnable
+            assertEquals(1, resourceTasks.size());
+            assertEquals(1, task.getResourceStats().size());
+            assertEquals(1, task.getResourceStats().get(threadId).size());
+            assertFalse(task.getResourceStats().get(threadId).get(0).isActive());
+
+            long expectedArrayAllocationOverhead = 2 * 4012688; // Task's memory overhead due to array allocations
+            long actualTaskMemoryOverhead = task.getTotalResourceStats().getMemoryInBytes();
+
+            assertTrue(actualTaskMemoryOverhead - expectedArrayAllocationOverhead < taskTestContext.memoryConsumptionWhenExecutionStarts);
+            assertTrue(task.getTotalResourceStats().getCpuTimeInNanos() > 0);
+        };
+
+        startResourceAwareNodesAction(testNodes[0], false, taskTestContext, new ActionListener<NodesResponse>() {
+            @Override
+            public void onResponse(NodesResponse listTasksResponse) {
+                responseReference.set(listTasksResponse);
+                taskTestContext.requestCompleteLatch.countDown();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                throwableReference.set(e);
+                taskTestContext.requestCompleteLatch.countDown();
+            }
+        });
+
+        // Waiting for whole request to complete and return successfully till client
+        taskTestContext.requestCompleteLatch.await();
+
+        assertTasksRequestFinishedSuccessfully(resourceTasks.size(), responseReference.get(), throwableReference.get());
+    }
+
+    public void testTaskResourceTrackingEnabledWhileTaskInProgress() throws Exception {
```

Comment:
Does the toggling of feature from on -> off -> on between the execution needs to be covered as well? I am assuming just for completeness as there are no guarantees for those tasks anyways.

Reply:
`on -> off -> on` will also remain the same as the logic uses the settings check only when task is received (We add such tasks to map). If someone disables setting while task is in progress then that has no impact on the running task and enabling it again will also not impact the running task. 

Given the above scenario I believe these 2 tests covers setting update impact scenarios.